### DISCUSSION TITLE
Why is my listing no longer showing up? Says my listing qmds contain no metadata therefore 'undefined'

### DISCUSSION THREAD

**Discussion by skycheng17** [quarto-dev/quarto-cli#8122] [Q&A] [answered]
### Description

This error seems to have only just come up so I wonder if it's either a bug or I amended my code somewhat without knowing.

I have a portfolio listing on the bottom my website homepage. It worked until today, where now the portfolio shows up empty. It is only empty when it is on my homepage and not on the designated portfolio tab.
<img width="1196" alt="Screenshot 2024-01-04 at 22 22 19" src="https://github.com/quarto-dev/quarto-cli/assets/100640344/ff790ca0-d1ee-4e2e-91a7-abeb2d2a0a2c">

This is the relevant code:
```
listing:

  - id: portfolio
    contents: "portfolio/**/index.qmd"
    type: table
    fields: [title, categories, date, image]
    sort: "date desc"
    sort-ui: false
    filter-ui: [title, date, categories]
    page-size: 3
    image-align: left
    image-height: 1.75in

---


## Portfolio

::: {#portfolio}
:::

#### [See more projects &rarr;](/portfolio)

```

On terminal, I keep getting this message:
```
WARNING: File /./././././docs/portfolio/6_Digital_Equity/index.qmd in the listing 'portfolio' contains no metadata.
```

How should I resolve this? It seems to work fine on the portfolio tab.

**Comment by mcanouil** [comment]
Could you share a small self-contained "working" (reproducible) example to work with, _i.e._, a Git repository? Thanks.
Your snippet is not enough.

What version of Quarto are you using?

**Reply by cscheid** [reply]
We need help from you, by you creating a _small_ self-contained reproducible example. It's hard to disentangle the causes on a large project like this.

**Reply by skycheng17** [reply]
Hm okay... with the small self-contained, reproducible snippet Git repository I've created, it seems to work.

https://github.com/skycheng17/test_portfolio

IDK why it doesn't work with my full website though. I don't know what's causing the empty metadata.

**Reply by skycheng17** [reply]
After adding a few more qmd objects to the listing, it now seems to be removing its own metadata during the preview--the listings show up briefly before deleting itself.

https://github.com/skycheng17/test_portfolio

Does this example work? Sorry for the confusion

**Reply by mcanouil** [reply]
Could you simplify it?
For instance, all the images, etc.
The simplest your example is the easiest it would be to figure out what is wrong. (Possibly a Quarto bug).

Also, you have a listing in a listing causing recursion in your project which is very likely the root cause of all the trouble here.

`contents: "portfolio/**/index.qmd"` can be equal to `contents: "portfolio/index.qmd"` meaning you recursively include the listing inside `portfolier`.
Move the "posts" inside `portfolio/contents`, then modify the paths accordingly.

**Reply by skycheng17** [reply]
Hi,

Thanks for the response. I'll test that out.

For now, I've got my website to work (skycheng17.github.io) . It did seem like a bug, as this issue of the undefined listing seemed to only occur when I was live previewing. However, after rendering, the issue stopped.

**Reply by mcanouil** [reply]
After fixing the recursive listing contents definition, preview worked just fine without any obvious changes abnormalities under the latest Quarto 1.4 pre-release.

**Reply by skycheng17** [reply]
Just to clarify, do you mean multiple index.qmds can be read in the listing, so each index.qmd and the rest of the contents per listing should be kept in the larger portfolio folder?

**Reply by mcanouil** [reply]
I figured the issue about the recurse thing (not the listings in fact):

```yml
resources: 
  - "portfolio/1_DigiPlay/index.qmd"
```

Why are you adding the Quarto document as a website resource? Do you want the Qmd to be uploaded with the rendered website?

I made a PR to cleanup your example.
Then all is good. (even with the qmd uploaded alongside the website because of setting resources)
<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/6aceab0f-a61a-48ef-a6d2-ac9d25bda3ba">

Then because `docs` is not hidden and is part of the project Quarto also looks into this directory and finds a Quarto document which it tries to render, and so on, and so on, at each rendering ...



https://github.com/quarto-dev/quarto-cli/assets/8896044/ebe14789-2a28-4d89-9a39-2ea2cd0f1bf9

Resulting in:

<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/2931f05d-f36f-4224-9d2a-e375f58307c4">

If you really intend to add Quarto documents to your website using `resources` or links, and more generally if you render to a directory that is not hidden to Quarto (starting with an underscore), then you need to ignore the directory:

```yaml
project:
  render:
    - "!docs"
```

Note that I am using Quarto 1.4 pre-release.

**Reply by mcanouil** [reply]
Hum, ignoring does not help as the listing is still seeing the file in `docs`.

Note that `quarto render` works, but not `quarto preview`:
| Render | Preview |
|--------|--------|
| <img width="1061" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/43020703-9afc-4c1d-b55a-10a5ace24d42"> | <img width="1063" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/1afc5906-82cb-4750-aa65-ddeee318eb3f"> | 

CC @dragonstyle @cscheid Here is the repository a bit cleaned up: https://github.com/mcanouil/test_portfolio/tree/chore/example-cleanup

**Reply by mcanouil** [reply, marked as answer]
@skycheng17 I really think there is a bug with Quarto Preview here and went ahead to open an issue:
- https://github.com/quarto-dev/quarto-cli/issues/8147

Note that if you want the Quarto document to be downloadable from your website, you could use `code-tools`: https://quarto.org/docs/output-formats/html-code.html#code-tools

Your listing is showing up by the way, it's just that a lot of bad content is added before your actual contents at each `quarto preview`.

**Reply by cscheid** [reply]
We did just fix a bug related to render lists yesterday. Are we positive that this still happens on `main`? I'm not following the details of the issue so if you could confirm ahead of me diving it that'd be great

**Reply by mcanouil** [reply]
I always tries on main ^^ in particular 094f50717a6854448c984ea0bd4b0b72fcfda260 for this one.

TL;DR: `quarto preview` seems to detect more than `quarto render` for listings which causes recursion in this particular case. This results in a "resource" document being including in the listing at each preview and the resource to be "uploaded" as well, etc.
| Tree | Warnings |
|--------|--------|
| <img width="355" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/58386efb-3209-4d05-846d-7c78af0e8171"> | <img width="1061" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/fcbef76d-8dd2-493c-8b29-403dab84a4a8"> | 

EDIT: `quarto preview --no-browser` does the right thing.
